pkg/controller/toolchainstatus/toolchainstatus_controller.go

```diff
@@ -0,0 +1,438 @@
+		toolchainStatus.Status.RegistrationService.RegistrationServiceResources.Conditions = []toolchainv1alpha1.Condition{*errCondition}
+		return err
+	}
+
+	// use the registrationservice resource directly in the toolchainstatus
+	toolchainStatus.Status.RegistrationService.RegistrationServiceResources.Conditions = registrationService.Status.Conditions
+	for _, condition := range registrationService.Status.Conditions {
+		if condition.Type == toolchainv1alpha1.ConditionReady && condition.Status != "True" {
+			return errs.Wrap(fmt.Errorf(condition.Message), "registration service resource not ready")
+		}
+	}
+	return nil
+}
+
+// handles the RegistrationService.Deployment part of the toolchainstatus
+func (s regServiceSubstatusHandler) addRegistrationServiceDeploymentStatus(reqLogger logr.Logger, toolchainStatus *toolchainv1alpha1.ToolchainStatus) error {
+	deploymentConditions, err := status.GetDeploymentStatusConditions(s.controllerClient, registrationservice.ResourceName, toolchainStatus.Namespace)
+	if err != nil {
+		err = errs.Wrap(err, "unable to determine the registration service deployment status")
+		errCondition := status.NewComponentErrorCondition(toolchainv1alpha1.ToolchainStatusReasonRegServiceDeploymentNotFound, err.Error())
+		toolchainStatus.Status.RegistrationService.Deployment.Conditions = []toolchainv1alpha1.Condition{*errCondition}
+		return err
+	}
+
+	toolchainStatus.Status.RegistrationService.Deployment.Name = registrationservice.ResourceName
+	toolchainStatus.Status.RegistrationService.Deployment.Conditions = deploymentConditions
+	return nil
+}
+
+// handles the RegistrationService.Health part of the toolchainstatus
+func (s regServiceSubstatusHandler) addRegistrationServiceHealthStatus(reqLogger logr.Logger, toolchainStatus *toolchainv1alpha1.ToolchainStatus) error {
+	// get the JSON payload from the health endpoint
+	registrationServiceHealthURL := "http://registration-service/" + registrationServiceHealthEndpoint
+	resp, err := s.httpClientImpl.Get(registrationServiceHealthURL)
+	if err != nil {
+		err = errs.Wrap(err, errMsgRegistrationServiceNotReady)
+		errCondition := status.NewComponentErrorCondition(toolchainv1alpha1.ToolchainStatusReasonRegServiceNotReady, err.Error())
+		toolchainStatus.Status.RegistrationService.Health.Conditions = []toolchainv1alpha1.Condition{*errCondition}
+		return err
+	}
+
+	// bad response
+	if resp.StatusCode != http.StatusOK || resp.Body == nil {
+		err = errs.Wrap(fmt.Errorf("bad response from registration service: %+v", resp), errMsgRegistrationServiceNotReady)
+		errCondition := status.NewComponentErrorCondition(toolchainv1alpha1.ToolchainStatusReasonRegServiceNotReady, err.Error())
+		toolchainStatus.Status.RegistrationService.Health.Conditions = []toolchainv1alpha1.Condition{*errCondition}
+		return err
+	}
+
+	// decode the response to JSON
+	defer resp.Body.Close()
```

Comment:
It will still leak FDs. You have to read the body first then close it: `ioutil.ReadAll(response.Body)`
See https://github.com/fabric8-services/fabric8-auth/blob/975831c9c2bb8a15ea8c8a0620008a4bb7d461a1/rest/rest.go#L140 as an example.

Reply:
Ah, didn't know that. Thanks for the tip. Handled it in 4fe2dd1087523aa210c2dea19fda210f0345bfb1